layers/ospf.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2017 Google, Inc. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file in the root of the source
+// tree.
+
+package layers
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/google/gopacket"
+)
+
+type OSPFType uint8
+
+// Potential values for OSPF.Type.
+const (
+	OSPFHello                   OSPFType = 1
+	OSPFDatabaseDescription     OSPFType = 2
+	OSPFLinkStateRequest        OSPFType = 3
+	OSPFLinkStateUpdate         OSPFType = 4
+	OSPFLinkStateAcknowledgment OSPFType = 5
+)
+
+func (i OSPFType) String() string {
+	switch i {
+	case OSPFHello:
+		return "Hello"
+	case OSPFDatabaseDescription:
+		return "Database Description"
+	case OSPFLinkStateRequest:
+		return "Link State Request"
+	case OSPFLinkStateUpdate:
+		return "Link State Update"
+	case OSPFLinkStateAcknowledgment:
+		return "Link State Acknowledgment"
+	default:
+		return ""
+	}
+}
+
+//OSPFv2 extend the OSPF head with version 2 specific fields
+type OSPFv2 struct {
+	AuType         uint16
+	Authentication uint64
+}
+
+// OSPFv3 extend the OSPF head with version 3 specific fields
+type OSPFv3 struct {
+	Instance uint8
+	Reserved uint8
+}
+
+// OSPF is a basic OSPF packet header with common fields of Version 2 and Version 3.
+type OSPF struct {
+	BaseLayer
+	OSPFv2
```

Comment:
Interesting use of embedded types here.  I'm not sure if I prefer your approach:

```go
type OSPF struct {
  OSPFv2
  OSPFv3
  ...
}
```

or this approach:

```go
type OSPF {
  ...  common fields ...
}

type OSPFv2 {
  OSPF
  ... v2 specific fields
}

type OSPFv3 {
  OSPF
  ... v3 specific fields ...
}
```

I'd be interested to hear your thoughts on each.  A question this brings up is whether an OSPFv2 packet should be a different layer type than an OSPFv3 one.